Firestore/core/src/firebase/firestore/model/transform_operations.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_TRANSFORM_OPERATIONS_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_TRANSFORM_OPERATIONS_H_
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+// TODO(zxu123): We might want to refactor transform_operations.h into several
```

Comment:
Can you provide some context on this class? As it stands, it appears that `ServerTimestampTransform` is the only member in the hierarchy, and `type` the only method. In other words, it doesn't appear to add much value over a simple `enum`, and it seems to only exist to allow an equivalent of a `switch` statement in `FSTMutation`. Are there any plans to expand the hierarchy?

Reply:
From what I read in the code, the `TransformOperation` is a protocal that is used everywhere, in particular in `FieldTransform` class, which is able to transform a particular field on a document. Right now, the only supported transform is `ServerTimestampTransform`, which just update the specified field with server-timestamp. A totally fabricated transform might be like `AdditionTransform`, which increase integer value by `X` for integer fields and ignore other fields, where `X` is specified when creating that `AdditionTransform` object. Right now the design does not allow transform multiple fields; but I can see potentially there could be `DocumentTransform` which deals with input being multiple fields (i.e. document).

However, you are right that for current code, we could just replace the `TransformOperation` and `ServerTimestampTransform` by an enum type, where ServerTimestamp is the only value (for now).

Reply:
FYI: Array transforms and Increment/Decrement handling is coming in Q2.